cliquenet/src/lib.rs

```diff
@@ -509,14 +584,20 @@ where
     /// This function will look up the x25519 public key of the ed25519 key
     /// and the remote address and then spawn a connection task.
     fn spawn_connect(&mut self, k: PublicKey) {
+        if self.connecting.contains_key(&k) {
+            debug!(node = %self.key, peer = %k, "connect task already started");
+            return;
+        }
         let x = self.index.get_by_left(&k).expect("known public key");
         let a = self.peers.get(&k).expect("known address");
-        self.connect_tasks.spawn(connect(
+        let h = self.connect_tasks.spawn(connect(
             (self.key, self.keypair.clone()),
             (k, *x),
             a.clone(),
             self.metrics.clone(),
         ));
+        assert!(self.task2key.insert(h.id(), k).is_none());
```

Comment:
Do we want to do this? Should we perhaps also log something descriptive if we do die?

Reply:
An invariant of this module is that adding to `task2key` only happens when no prior mapping for the task ID exists. It would be a programmer error if it does, so I think the assertion expresses it well. Would you prefer this slightly longer variant?

```rust
let prev = self.task2key.insert(h.id(), k);
assert!(prev.is_none(), "task to key mapping already exists");
```